From f4e5d7088dad31ddba3ef4496db1c19bb06f4bca Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:24:50 -0400
Subject: [tools] perf/symbols: Add new option --ignore-vmlinux for perf top

Message-id: <1408462094-14194-122-git-send-email-jolsa@redhat.com>
Patchwork-id: 88182
O-Subject: [PATCH RHEL7.1 BZ1131394 121/325] perf symbols: Add new option --ignore-vmlinux for perf top
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit fc2be6968e99b5314f20e938a547d44dcb1c40eb
Author: Willy Tarreau <w@1wt.eu>
Date: Sat Sep 14 10:32:59 2013 +0200

description
===========
Running "perf top" on a machine with possibly invalid or non-matching
vmlinux at the various places results in no symbol resolving despite
/proc/kallsyms being present and valid.

Add a new option --ignore-vmlinux to explicitly indicate that we do not
want to use these kernels and just use what we have (kallsyms).
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-top.c b/tools/perf/builtin-top.c
index 5986c54..7afbe64 100644
--- a/tools/perf/builtin-top.c
+++ b/tools/perf/builtin-top.c
@@ -1075,6 +1075,8 @@ int cmd_top(int argc, const char **argv, const char *prefix __maybe_unused)
       "list of cpus to monitor"),
  OPT_STRING('k', "vmlinux", &symbol_conf.vmlinux_name,
      "file", "vmlinux pathname"),
+ OPT_BOOLEAN(0, "ignore-vmlinux", &symbol_conf.ignore_vmlinux,
+      "don't load vmlinux even if found"),
  OPT_BOOLEAN('K', "hide_kernel_symbols", &top.hide_kernel_symbols,
       "hide kernel symbols"),
  OPT_CALLBACK('m', "mmap-pages", &opts->mmap_pages, "pages",
diff --git a/tools/perf/util/symbol.c b/tools/perf/util/symbol.c
index cd1dcc4..48c3879 100644
--- a/tools/perf/util/symbol.c
+++ b/tools/perf/util/symbol.c
@@ -1215,7 +1215,7 @@ static int dso__load_kernel_sym(struct dso *dso, struct map *map,
   goto do_kallsyms;
  }
 
- if (symbol_conf.vmlinux_name != NULL) {
+ if (!symbol_conf.ignore_vmlinux && symbol_conf.vmlinux_name != NULL) {
   err = dso__load_vmlinux(dso, map,
      symbol_conf.vmlinux_name, filter);
   if (err > 0) {
@@ -1227,7 +1227,7 @@ static int dso__load_kernel_sym(struct dso *dso, struct map *map,
   return err;
  }
 
- if (vmlinux_path != NULL) {
+ if (!symbol_conf.ignore_vmlinux && vmlinux_path != NULL) {
   err = dso__load_vmlinux_path(dso, map, filter);
   if (err > 0)
    return err;
diff --git a/tools/perf/util/symbol.h b/tools/perf/util/symbol.h
index 2a97bb1..9b8b213 100644
--- a/tools/perf/util/symbol.h
+++ b/tools/perf/util/symbol.h
@@ -85,6 +85,7 @@ struct symbol_conf {
  unsigned short priv_size;
  unsigned short nr_events;
  bool  try_vmlinux_path,
+   ignore_vmlinux,
    show_kernel_path,
    use_modules,
    sort_by_name,
-- 
1.7.1